contracts/registrar/contract/registrar.sol

```diff
@@ -0,0 +1,272 @@
+        uint count; // Number of signers who have submitted checkpoint announcement
+        mapping(address => bytes32) usermap; // map between signer address and advertised checkpoint hash
+        mapping(bytes32 => Vote[]) votemap; // map between checkpoint hash and relative signer announcements.
+    }
+
+    /*
+        Modifiers
+    */
+
+    /**
+     * @dev Check whether the message sender is authorized.
+     */
+    modifier OnlyAuthorized() {
+        require(admins[msg.sender]);
+        _;
+    }
+
+    /*
+        Events
+    */
+
+    // NewCheckpointEvent is emitted when a new checkpoint proposal receive enough approvals.
+    // We use checkpoint hash instead of the full checkpoint to make the transaction cheaper.
+    event NewCheckpointEvent(uint indexed index, bytes32 checkpointHash, bytes signature);
+
+    /*
+        Public Functions
+    */
+    constructor(address[] memory _adminlist, uint _sectionSize, uint _processConfirms, uint _threshold) public {
+        for (uint i = 0; i < _adminlist.length; i++) {
+            admins[_adminlist[i]] = true;
+            adminList.push(_adminlist[i]);
+        }
+        sectionSize = _sectionSize;
+        processConfirms = _processConfirms;
+        threshold = _threshold;
+    }
+
+    /**
+     * @dev Get latest stable checkpoint information.
+     * @return section index
+     * @return checkpoint hash
+     * @return block height associated with checkpoint
+     */
+    function GetLatestCheckpoint()
+    view
+    public
+    returns(uint, bytes32, uint) {
+        (bytes32 hash, uint height) = GetCheckpoint(latest);
+        return (latest, hash, height);
+    }
+
+    /**
+     * @dev Get a stable checkpoint information with specified section index.
+     * @param _sectionIndex section index
+     * @return checkpoint hash
+     * @return the associated register block height
+     */
+    function GetCheckpoint(uint _sectionIndex)
+    view
+    public
+    returns(bytes32, uint)
+    {
+        return (checkpoints[_sectionIndex], register_height[_sectionIndex]);
+    }
+
+    /**
+     * @dev Submit a new checkpoint announcement
+     * Checkpoint represents a set of post-processed trie roots (CHT and BloomTrie)
+     * associated with the appropriate section head hash.
+     *
+     * It is used to start light syncing from this checkpoint and avoid downloading
+     * the entire header chain while still being able to securely access old headers/logs.
+     *
+     * @param _sectionIndex section index
+     * @param _hash checkpoint hash calculated in the client side
+     * @param _sig admin's signature for checkpoint hash
+     *         `checkpoint_hash = Hash(index, sectionHead, chtRoot, bloomRoot)`
+     *         `_sig = Sign(privateKey, checkpoint_hash)`
+     * @return indicator whether set checkpoint successfully
+     */
+    function SetCheckpoint(
+        uint _sectionIndex,
+        bytes32 _hash,
+        bytes memory _sig
```

Comment:
We had this idea for reorg protection: When voting, also pass in the current block number and hash (or maybe one from 10 blocks). That way if there's a reorg, the votes will fail on that "chain" and we don't serve up bad CHTs, rather revert to an older set one.